### ISSUE TITLE
Ideal reward window and number of addresses per commit

### ISSUE THREAD

**Issue by jeffdomke** [blockstack/stacks-blockchain#1847]
Current design with 1000 blocks, with five outputs per block, will incur ~30% more Bitcoin transaction fees vs. a 5000 block design. Which is superior and why? 

1000 block
- Stacks locked for shorter time

5000 block
- Stacks locked for longer time
- Everyone pays ~30% fewer Bitcoin fees
- Seemingly more attractive for Stackers with smaller holdings / encourages more delegation

@jessesoslow curious if your model can add color here? 

@psq thx for the fee information, feel free to add more details.

**Comment by diwakergupta** [comment]
Couple of points that were brought up in the meeting today for additional context:

- once we address #1615 , we'll already gain substantial savings in fee (~40% I believe)
- @kantai had brought up possible scenarios with a single output where a miner could effectively ensure it wins a "free" block

Extrapolating this line of thinking further, why stop at 5000 blocks. If these arguments hold, why not 10K blocks (single output every other block) or higher?

I suspect it's impossible for us to come up with a provably optimal window a priori. As with other choices around PoX, I'd optimize for building safe-guards and optionality, rather than trying to come up with a correct "magic" number -- we'll most likely get it wrong no matter what.

**Comment by jcnelson** [comment]
> Seemingly more attractive for Stackers with smaller holdings / encourages more delegation

Not sure how this is a point in favor of the 5000 block proposal.  Delegates make it so _anyone_ can Stack, no matter how few STX they hold, and no matter how long the reward cycle is (the amount of STX to lock-up for a reward address is is the same regardless).

**Comment by jessesoslow** [comment]
I agree with @jcnelson -- not clear to me how this change would impact delegating and/or attractiveness to smaller holders.  (Other then, as with all participants, less fees presumably means more value being transferred to Stackers instead of BTC miners.)

> Extrapolating this line of thinking further, why stop at 5000 blocks. If these arguments hold, why not 10K blocks (single output every other block) or higher?
> 
> I suspect it's impossible for us to come up with a provably optimal window a priori. As with other choices around PoX, I'd optimize for building safe-guards and optionality, rather than trying to come up with a correct "magic" number -- we'll most likely get it wrong no matter what.

I both agree and disagree with @diwakergupta here.  I think finding the provably optimal window a priori is not the right goal.  Instead, I think a fair question, is whether the benefits of distributing to 5 addresses each block is worth the added fees?  I think _if_ we're still planning on implementing a penalizing smoothing function, in the form of the average of the median and the minimum for the previous 5 (or X) blocks, then it's not obvious to me how a miner could ensure they win a free block (by say contributing a large sum of BTC in a block with their own address) without suffering significant cost in multiple other blocks.  I would think the only way that wouldn't be the case is if the miner also had multiple addresses in the surrounding blocks.  But this would be roughly equivalent, probabilistically, to the miner having all (or most) of the 5 addresses in the 5 address block, were we to stick with the original approach.  But I'm sure @kantai has a deeper understanding of this than I do, so don't trust me here.

> @jessesoslow curious if your model can add color here?

I'm not sure the model has a ton to add here, though we're digging into the model with the person working on it today.  So that could lead to some additional thoughts.

I think the primary takeaway is that we'd expect with this change that value to Stackers would increase by an amount >0 and ≤~.3*(Total BTC Fees), which is generally a good a thing.  But if @kantai is right that this leads to scenarios where miners can effectively mine for free, then it's surely not worth it.

> once we address #1615 , we'll already gain substantial savings in fee (~40% I believe)

Lastly, I think the above change is great.  But if we can improve it further with little downside, I think we should.  As mentioned, the extent of the downside isn't clear to me, though.

**Comment by jcnelson** [comment]
> I think if we're still planning on implementing a penalizing smoothing function, in the form of the average of the median and the minimum for the previous 5 (or X) blocks, then it's not obvious to me how a miner could ensure they win a free block (by say contributing a large sum of BTC in a block with their own address) without suffering significant cost in multiple other blocks. @jessesoslow 

We are still planning to do this.

**Comment by kantai** [comment]
>  But this would be roughly equivalent, probabilistically, to the miner having all (or most) of the 5 addresses in the 5 address block, were we to stick with the original approach. But I'm sure @kantai has a deeper understanding of this than I do, so don't trust me here.

Yep, I agree: with a penalizing smoothing function, the automatic "free block" problem in the 1-address scheme would be mitigated by the smoothing function.

> I think the primary takeaway is that we'd expect with this change that value to Stackers would increase by an amount >0 and ≤~.3*(Total BTC Fees), which is generally a good a thing.

I'm not yet convinced that's the case. _Miners_ would face lower minimum participation thresholds for STX mining, but Stackers might not reap these benefits, because there's an amortization of fixed costs that happens when the number of output addresses increases (because reducing addresses from 5 to 1 only yields a 30% reduction in Bitcoin fees).

Think of it in terms of Bitcoin fees paid _over the length of the reward cycle_:

Let `F := transaction fee for a 5 address transaction`

1. In the "1-address case", there would be 5,000 transactions, so the total paid bitcoin fee is:

```
5000 * (.7F) = 3500F
```

2. In the "5-address case", there would be 1,000 transactions, so the total paid bitcoin fee is:

```
1000 * (F) = 1000F
```

**Comment by jessesoslow** [comment]
> Think of it in terms of Bitcoin fees paid over the length of the reward cycle:

Isn't the right normalizing frame a set number of blocks, rather than differing reward cycle lengths?  I would think it would make more sense to compare, say, 5000 blocks under both scenarios, rather than the reward cycles against each other.

Let's assume 5k blocks under both scenarios.  Let's say the total amount of BTC transferred to addresses by miners during these 5k blocks is M and the total fees paid by these miners is F.  Let's say in the 5-address case it's M_0 and F_0, and in the 1-address case it's M_1 and F_1.

I believe it's reasonable to assume that over the same number of blocks M_0 + F_0 ≈ M_1 + F_1.  In other words, the total amount transferred by miners, including fees, for the same number of STX should be approximately equivalent in either case.

I believe the fees over this period you would have the same number of transactions, but in the 1-address case, the fees for those transactions should be about 30% less.  I believe that would mean F_1 ≈ (.7)*F_0

And so M_0 + F_0 ≈ M_1 + (.7)*F_0 --> M_0 + (.3)F_0 ≈ M_1

**Comment by kantai** [comment]
I agree with that, @jessesoslow, and I'm convinced that a reduction in number of addresses would likely increase the value captured by Stackers.

**Comment by psq** [comment]
> I believe it's reasonable to assume that over the same number of blocks M_0 + F_0 ≈ M_1 + F_1. In other words, the total amount transferred by miners, including fees, for the same number of STX should be approximately equivalent in either case.

This assumes the number of miners is constant, or near constant, which I doubt will be the case.  Lower fees will primarily mean more miners, which is desirable anyway.  A miner has no incentive to favor `M` over `F`, `M+F` is what matters to miners. My intuition tells me that `F`, being unavoidable, will be the main driver, and that the number of miners will increase to keep `F` close to `reward`, reducing/keeping `M` to the remainder (that's assuming miners will want to mine profitably, and this does not account for rejected blocks, which lowers the rewards).

So if:
```
F = m * f

m: number of miners
f: fee per miner
```
we'll have:
```
F1 ≈ F0
```
which means
```
m1 * (f * (1 - 0.3)) = m0 * f
m1 = m0 * 1/0.7 = 42% more miners
```

**Comment by jessesoslow** [comment]
@psq I think you make a good point, although I think reality will likely reflect some middle ground between 30% more value being distributed to Stackers and 42% more miners.

In particular, I believe your calculation essentially assumes all value flows to fees.  Correct me if I'm wrong, but I believe the corollary of that premise is that in equilibrium, each miner's total contribution would be equal to f + ε, where ε is a nominal amount.  I don't think that's what an equilibrium would look like, though.

If we imagine that's the case, then what's the behavior of a miner?  If there are n miners then the total amount being distributed would be n * ε, which given the smallness of ε would also by definition be small (and well less than the value of rewards (n * F less, to be precise)).  A miner need only contribute a small amount more to drastically increase their chances of winning.

To take a super simplified example with made up numbers, assume you have a rewards block that's worth 1050 satoshis and that fees are 20 satoshis per transaction. If we assume that, in equilibrium, the number of miners will ultimately be maximized, then the outcome you'd expect would be 50 miners contributing 1 satoshi as m and 20 satoshis as f.  In that case, every miner has an expected return of 0.

But if any single miner increases their m to 2 satoshis, they increase their expected return to 87% ((2 / 51) * 1050 = 41.2 and (41.2 - 22) / 22 = .87).  If a single miner increases m to 30 satoshis, their expected return becomes ~700%.  Clearly that state isn't stable.

Of course my example is extraordinarily over-simplified both in it representation of the system and actions of participants. Nevertheless, I think the example is illustrative and is related to the statement that

> A miner has no incentive to favor M over F

which I don't think is quite correct.  The reason to favor M over F is that M impacts probability of winning and F doesn't.

Regardless of all this, I think we both agree that -- at least in a vacuum -- less fees is better.  

I'm curious if there are downsides to moving to 1 address from 5 that we haven't thought about?

**Comment by diwakergupta** [comment]
> I'm curious if there are downsides to moving to 1 address from 5 that we haven't thought about?

In the interest of bringing this issue to a closure, any objections to decreasing the per-block PoX reward addresses? If not, we should finalize the new value, assuming we'll fallback on 1 address.

I'm still worried about the user-experience with a 5000 block cycle (~5 weeks) with 1 address. Would we consider 2 addresses? It would bring the reward cycle down to 2.5 weeks while still preserving many of the benefits. Any product input @jeffdomke ?

**Comment by jcnelson** [comment]
> In the interest of bringing this issue to a closure, any objections to decreasing the per-block PoX reward addresses?

I think it should remain as-is, precisely because of the impact on the user experience.  Why force users to wait a month for PoX rewards, when it could be a week?  I know @muneeb-ali has repeatedly emphasized this point -- if users have to wait a month before they see anything for Stacking (which itself will likely be a novel, tedious experience for them), they may not think it's worth it to even keep their STX.

**Comment by kantai** [comment]
I'm still not sure if 1-address versus 5-addresses is preferable, but one potential benefit to a 1-address scheme is that it ensures that all the commits have roughly the same minimum transaction bytes: 1 OPRETURN and BTC-commit output address. In the 5-address scheme, this would mean that miners who are _not_ building off of the PoX anchor block (and therefore are attempting a deep reorg), would be advantaged over other miners (because they are paying ~30% less in BTC fees).

**Comment by jcnelson** [comment]
That's a good point.  If we go with anything more than 1 BTC output per block-commit, we should probably require that all block-commit transactions at least be the same size to avoid this outcome.

**Comment by kantai** [comment]
Circling back on this discussion-- I think the conclusion we reached is that we're going to move forward with _one_ reward address per commitment, which results in a reward cycle length of 5000 blocks (as opposed to 1000 blocks).

**Comment by jcnelson** [comment]
A 5000-block reward cycle is a long time -- it's almost 5 weeks.  It also means that the maximum lock-up period for STX (12 reward cycles) is now about 15 months.  Is this all good by everyone?

**Comment by muneeb-ali** [comment]
I'm inclined towards shorter cycles. Mainly because people who want to keep STX locked for longer (say months) can do so with shorter cycles but the other option of a minimum 5 weeks cycle feels like taking away optionality from people.

I understand the fee issue and think that is important as well.

**Comment by jcnelson** [comment]
I think we have consensus right now on:

* 2 reward addresses per block
* 2000-block reward cycle

I'll leave this open until this is reflected in SIP-007.

**Comment by kantai** [comment]
Closed by #1975
